pkg/detector/library/advisory.go

```diff
@@ -0,0 +1,102 @@
+package library
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/Masterminds/semver/v3"
+	"golang.org/x/xerrors"
+
+	"github.com/aquasecurity/trivy-db/pkg/db"
+	dbTypes "github.com/aquasecurity/trivy-db/pkg/types"
+	"github.com/aquasecurity/trivy/pkg/scanner/utils"
+	"github.com/aquasecurity/trivy/pkg/types"
+)
+
+// Advisory represents security advisories for each programming language
+type Advisory struct {
+	lang     string
+	comparer comparer
+}
+
+func NewAdvisory(lang string) *Advisory {
+	return &Advisory{
+		lang:     lang,
+		comparer: newComparer(lang),
+	}
+}
+
+// DetectVulnerabilities scans buckets with the prefix according to the programming language in "Advisory".
+// If "lang" is python, it looks for buckets with "python::" and gets security advisories from those buckets.
+// It allows us to add a new data source with the lang prefix (e.g. python::new-data-source)
+// and detect vulnerabilities without specifying a specific bucket name.
+func (s *Advisory) DetectVulnerabilities(pkgName string, pkgVer *semver.Version) ([]types.DetectedVulnerability, error) {
+	// e.g. "python::"
+	prefix := fmt.Sprintf("%s::", s.lang)
+	advisories, err := db.Config{}.GetAdvisories(prefix, pkgName)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to get %s advisories: %w", s.lang, err)
+	}
+
+	var vulns []types.DetectedVulnerability
+	for _, advisory := range advisories {
+		if !s.comparer.isVulnerable(pkgVer, advisory) {
+			continue
+		}
+
+		vuln := types.DetectedVulnerability{
+			VulnerabilityID:  advisory.VulnerabilityID,
+			PkgName:          pkgName,
+			InstalledVersion: pkgVer.String(),
+			FixedVersion:     s.createFixedVersions(advisory),
+		}
+		vulns = append(vulns, vuln)
+	}
+
+	return vulns, nil
+}
+
+func (s *Advisory) createFixedVersions(advisory dbTypes.Advisory) string {
+	if len(advisory.PatchedVersions) != 0 {
+		return strings.Join(advisory.PatchedVersions, ", ")
+	}
+
+	var fixedVersions []string
+	for _, version := range advisory.VulnerableVersions {
+		for _, s := range strings.Split(version, ",") {
+			s = strings.TrimSpace(s)
+			if !strings.HasPrefix(s, "<=") && strings.HasPrefix(s, "<") {
+				s = strings.TrimPrefix(s, "<")
+				fixedVersions = append(fixedVersions, strings.TrimSpace(s))
+			}
+		}
+	}
+	return strings.Join(fixedVersions, ", ")
+}
+
+type comparer interface {
+	isVulnerable(pkgVer *semver.Version, advisory dbTypes.Advisory) bool
+}
+
+func newComparer(lang string) comparer {
+	switch lang {
+	case "java":
+		// TODO: we need to use another library for version comparison in the case of Java.
+	}
```

Comment:
Should we take it out for now?

Reply:
When someone will work on Java support, I thought this comment would be helpful since he can know where he needs to add the logic. But we can remove it for now.

Reply:
done